[New] add `no-invalid-html-attribute` rule

Nokel81 opened the pull request:
This rule (and fixer) checks the "rel" value to make sure that it correctly relates to the tag that the "rel" property is on.

golopot commented on lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,190 @@
+    return {
+      JSXAttribute(node) {
+        // ignore attributes that aren't "rel"
+        if (node.type !== 'JSXIdentifier' && node.name.name !== 'rel') {
```
`node.type` is always `JSXAttribute` here.

golopot commented on tests/lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,464 @@
+    {code: '<a rel={{a: "noreferrer"}["a"]}></a>'},
+    {code: '<a rel={{a: "noreferrer"}["b"]}></a>'}
+  ],
+  invalid: [
```
missing tests for `<a rel="" />` and `<any rel />`

golopot commented on lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,190 @@
+        if (node.value.type === 'Literal') {
+          return checkLiteralValueNode(context, node.value, parentNodeName);
+        }
+
```
```suggestion
         if (node.value.type !== 'JSXExpressionContainer') {
          return;
        }
```
For readability it is nice to always assert node types.

golopot commented on tests/lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,464 @@
+const ruleTester = new RuleTester({parserOptions});
+
+ruleTester.run('jsx-no-invalid-rel', rule, {
+  valid: [
```
needs tests for non-html components like `<Foo rel="abc">`.

golopot commented on lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,190 @@
+  if (typeof node.value !== 'string') {
+    return context.report({
+      node,
+      messageId: 'onlyStrings',
```
tests are failing because eslint@4 does not support messageId and template messages.

ljharb commented on docs/rules/jsx-no-invalid-rel.md:
```diff
@@ -0,0 +1,14 @@
+# Prevent usage of invalid `rel` (react/jsx-no-invalid-rel)
+
+The JSX elements: `a`, `area`, `link`, or `form` all have a attribute called `rel`. There is is fixed list of values that have any meaning on these tags (see [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel)). To help with minimizing confusion while reading code, only the appropriate values should be on each attribute.
```
```suggestion
The JSX elements: `a`, `area`, `link`, or `form` all have an attribute called `rel`. There is is fixed list of values that have any meaning on these tags (see [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel)). To help with minimizing confusion while reading code, only the appropriate values should be on each attribute.
```

ljharb commented on tests/lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,464 @@
+    },
+    {
+      code: '<area rel="manifest"></area>',
+      output: '<area rel=""></area>',
```
seems like it should just remove the `rel` value entirely?

Nokel81 replied:
The eslint docs state that if any fixes are applied then a subsequent linting cycle is done. I believe that that would fix this.

ljharb replied:
While that's true, it seems ideal to avoid forcing an unnecessary fix pass, especially since in editor integrations, only one pass is run. I'd prefer to optimize this.

ljharb reviewed (changes requested):
Plus all of @golopot's comments.

ljharb commented on docs/rules/jsx-no-invalid-rel.md:
```diff
@@ -0,0 +1,16 @@
+# Prevent usage of invalid `rel` (react/jsx-no-invalid-rel)
+
+The JSX elements: `a`, `area`, `link`, or `form` all have an attribute called `rel`.
+There is is fixed list of values that have any meaning on these tags (see [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel)).
```
```suggestion
There is a fixed list of values that have any meaning for this attribute on these tags (see [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel)).
```

ljharb commented on lib/rules/jsx-no-invalid-rel.js:
```diff
@@ -0,0 +1,192 @@
+          return context.report({
+            node,
+            message: 'The "rel" attribute only has meaning on `<link>`, `<a>`, `<area>`, and `<form>` tags.',
+            fix(fixer) {
+              return fixer.remove(node);
+            }
```
this fixer function can be defined at module level and reused in a bunch of places.

ljharb reviewed (changes requested):
Are there other HTML element attributes that this type of rule might work on?

It might be nicer to make this rule a generic `no-invalid-html-attribute` rule (one that also works on `React.createElement`, and thus isn't jsx-specific), with an option for a list of attributes to check (that would default to `['rel']`). That way, we could add support for a new attribute in a semver-minor without needing to add an entirely new rule.

Thoughts?

Nokel81 commented:
@ljharb Good idea about making this a more generic rule, though I don't know if I want to make it working with `React.createElement` in this PR.

Nokel81 commented:
@ljharb I have changed the name of the rule and added the option as you said. While reworking the internals a bit (though I haven't tested the code with non-rel items)

ljharb commented on docs/rules/no-invalid-html-attribute.md:
```diff
@@ -0,0 +1,17 @@
+# Prevent usage of invalid `rel` (react/no-invalid-html-attribute)
```
i think the docs here still need updating?

ljharb commented on lib/rules/no-invalid-html-attribute.js:
```diff
@@ -0,0 +1,219 @@
+ * @type {Map<string, Map<string, Set<string>>>}
+ */
+const VALID_VALUES = new Map();
+VALID_VALUES.set('rel', new Map())
+  .get('rel')
+  .set('alternate', new Set(['link', 'area', 'a']))
+  .set('author', new Set(['link', 'area', 'a']))
+  .set('bookmark', new Set(['area', 'a']))
+  .set('canonical', new Set(['link']))
+  .set('dns-prefetch', new Set(['link']))
+  .set('external', new Set(['area', 'a', 'form']))
+  .set('help', new Set(['link', 'area', 'a', 'form']))
+  .set('icon', new Set(['link']))
+  .set('license', new Set(['link', 'area', 'a', 'form']))
+  .set('manifest', new Set(['link']))
+  .set('modulepreload', new Set(['link']))
+  .set('next', new Set(['link', 'area', 'a', 'form']))
+  .set('nofollow', new Set(['area', 'a', 'form']))
+  .set('noopener', new Set(['area', 'a', 'form']))
+  .set('noreferrer', new Set(['area', 'a', 'form']))
+  .set('opener', new Set(['area', 'a', 'form']))
+  .set('pingback', new Set(['link']))
+  .set('preconnect', new Set(['link']))
+  .set('prefetch', new Set(['link']))
+  .set('preload', new Set(['link']))
+  .set('prerender', new Set(['link']))
+  .set('prev', new Set(['link', 'area', 'a', 'form']))
+  .set('search', new Set(['link', 'area', 'a', 'form']))
+  .set('stylesheet', new Set(['link']))
+  .set('tag', new Set(['area', 'a']));
```
```suggestion
const rel = new Map([
  ['alternate', new Set(['link', 'area', 'a'])],
  ['author', new Set(['link', 'area', 'a'])],
  ['bookmark', new Set(['area', 'a'])],
  ['canonical', new Set(['link'])],
  ['dns-prefetch', new Set(['link'])],
  ['external', new Set(['area', 'a', 'form'])],
  ['help', new Set(['link', 'area', 'a', 'form'])],
  ['icon', new Set(['link'])],
  ['license', new Set(['link', 'area', 'a', 'form'])],
  ['manifest', new Set(['link'])],
  ['modulepreload', new Set(['link'])],
  ['next', new Set(['link', 'area', 'a', 'form'])],
  ['nofollow', new Set(['area', 'a', 'form'])],
  ['noopener', new Set(['area', 'a', 'form'])],
  ['noreferrer', new Set(['area', 'a', 'form'])],
  ['opener', new Set(['area', 'a', 'form'])],
  ['pingback', new Set(['link'])],
  ['preconnect', new Set(['link'])],
  ['prefetch', new Set(['link'])],
  ['preload', new Set(['link'])],
  ['prerender', new Set(['link'])],
  ['prev', new Set(['link', 'area', 'a', 'form'])],
  ['search', new Set(['link', 'area', 'a', 'form'])],
  ['stylesheet', new Set(['link'])],
  ['tag', new Set(['area', 'a'])],
]);
VALID_VALUES.set('rel', rel);
```

ljharb commented on lib/rules/no-invalid-html-attribute.js:
```diff
@@ -0,0 +1,219 @@
+function splitIntoRangedParts(node, regex) {
+  const res = [];
+  const valueRangeStart = node.range[0] + 1; // the plus one is for the initial quote
+  let match;
+
+  // eslint-disable-next-line no-cond-assign
+  while ((match = regex.exec(node.value)) !== null) {
```
please use https://npmjs.com/string.prototype.matchall and `Array.from` for this rather than a while loop.

ljharb commented on lib/rules/no-invalid-html-attribute.js:
```diff
@@ -0,0 +1,219 @@
+
+  const parentNodeName = node.parent.name.name;
+  if (!COMPONENT_ATTRIBUTE_MAP.has(attribute) || !COMPONENT_ATTRIBUTE_MAP.get(attribute).has(parentNodeName)) {
+    const tagNames = Array.from(COMPONENT_ATTRIBUTE_MAP.get(attribute).values())
+      .map((tagName) => `"<${tagName}>"`)
+      .join(', ');
```
```suggestion
    const tagNames = Array.from(
      COMPONENT_ATTRIBUTE_MAP.get(attribute).values(),
      (tagName) => `"<${tagName}>"`
    ).join(', ');
```

ljharb reviewed (changes requested):
Thanks, this is looking much better

Nokel81 commented:
I know that there are some failing tests for some older versions of eslint. Will look at them soon.

ljharb commented:
No rush, please mark as ready for review once tests are passing.

ljharb reviewed (approved):
Thanks, this looks great! (sorry for the review delay)

This rule isn't named "jsx-", but it doesn't check `React.createElement`. I think it probably should, though. Is that something you'd mind adding?

Nokel81 commented:
@ljharb I have added checking `React.createElement`.

ljharb reviewed (approved):
Are there other attributes besides "rel" we can add here?

Nokel81 commented:
Maybe https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/autocomplete but I think that could easily be done in a separate PR.

ljharb commented:
Rebased again; I tweaked the test. `<Foo rel="" />` must always be valid - this rule should only be checking HTML elements, never custom elements.

A few tests are still failing. @Nokel81?

Nokel81 commented:
Well since you just changed the logic or tests (cannot tell). I was whitelisting elements whereas I think you want me to blacklist html elements that aren't on the list.

ljharb commented:
All I did is move the two "Foo" tests from invalid to valid.

ljharb commented:
I'm fine with an inclusion-list-based approach, but it should ignore custom elements entirely.

Nokel81 commented:
@ljharb Sorry it took so long, this PR has been updated to ignore non-HTML tags.

ljharb commented:
@Nokel81 i enabled all the test cases and used the messageId pattern and rebased.
